src/core/tsi/alts/handshaker/alts_handshaker_client.cc

```diff
@@ -358,6 +335,164 @@ static tsi_result make_grpc_call(alts_handshaker_client* c, bool is_start) {
+    arg->client = client;
+    arg->self = this;
+    GRPC_CLOSURE_INIT(&arg->closure, EnqueueHandshakeAndMaybeStartSomeLocked,
+                      arg, nullptr);
+    combiner_->Run(&arg->closure, GRPC_ERROR_NONE);
+  }
+
+ private:
+  struct Node {
+    alts_grpc_handshaker_client* client = nullptr;
+    Node* next = nullptr;
+    Node* prev = nullptr;
+  };
+
+  struct Arg {
+    alts_grpc_handshaker_client* client;
+    HandshakeQueue* self;
+    grpc_closure closure;
+  };
+
+  static void EnqueueHandshakeAndMaybeStartSomeLocked(
+      void* arg, grpc_error* unused_error) {
+    Arg* args = static_cast<Arg*>(arg);
+    args->self->EnqueueHandshakeAndMaybeStartSomeInnerLocked(args->client);
+    grpc_core::Delete(args);
+  }
+
+  void EnqueueHandshakeAndMaybeStartSomeInnerLocked(
+      alts_grpc_handshaker_client* client) {
+    if (client == nullptr) {
+      outstanding_handshakes_--;
+    } else {
+      Node* node = grpc_core::New<Node>();
+      node->client = client;
+      gpr_log(GPR_DEBUG,
+              "HandshakeQueue:%p: outstanding_handshakes_:%ld "
+              "old head_:%p new head_: %p tail_:%p",
+              this, outstanding_handshakes_, head_, node, tail_);
+      if (head_ == nullptr) {
+        GPR_ASSERT(tail_ == nullptr);
+        head_ = node;
+        tail_ = node;
+      } else {
+        GPR_ASSERT(tail_ != nullptr);
+        node->next = head_;
+        GPR_ASSERT(head_->prev == nullptr);
+        head_->prev = node;
+        head_ = node;
+      }
+    }
+    while (outstanding_handshakes_ < max_outstanding_handshakes_ &&
+           tail_ != nullptr) {
+      gpr_log(GPR_DEBUG,
+              "start handshake "
+              "outstanding_handshakes_:%ld "
+              "head_:%p tail_:%p",
+              outstanding_handshakes_, head_, tail_);
+      GPR_ASSERT(outstanding_handshakes_ < max_outstanding_handshakes_);
+      continue_make_grpc_call(tail_->client, true /* is_start */);
+      Node* prev = tail_;
+      tail_ = tail_->prev;
+      grpc_core::Delete(prev);
+      outstanding_handshakes_++;
+    }
+    if (tail_ == nullptr) {
+      head_ = nullptr;
+    }
+  }
+
+  grpc_core::Combiner* combiner_ = grpc_combiner_create();
+  Node* head_ = nullptr;
+  Node* tail_ = nullptr;
+  size_t outstanding_handshakes_ = 0;
+  const size_t max_outstanding_handshakes_;
+};
+
+gpr_once g_queued_handshakes_init = GPR_ONCE_INIT;
+/* Using separate queues for client and server handshakes is a
+ * hack that's mainly intended to satisfy the alts_concurrent_connectivity_test,
+ * which runs many concurrent handshakes where both endpoints
+ * are in the same process; this situation is problematic with a
+ * single queue because we have a high chance of using up all outstanding
+ * slots in the queue, such that there aren't any
+ * mutual client/server handshakes outstanding at the same time and
+ * able to make progress. */
+HandshakeQueue* g_client_handshake_queue;
+HandshakeQueue* g_server_handshake_queue;
```

Comment:
It seems weird to keep an unused data structure in non-test code. If we increase the queue size in the test, can the problem be solved?

Reply:
Note that this data structure <i>is</i> used in the real code (i.e. outside the test) - `g_client_handshake_queue` is used by client side handshakes and `g_server_handshake_queue` is used by server side handshakes.

The problem being addressed here is basically the same problem that https://github.com/grpc/grpc/blob/970f703337ecc19017a737cc4eb831cd50fd0fec/test/core/tsi/alts/handshaker/alts_concurrent_connectivity_test.cc#L180 currently addresses in the test - so the workaround isn't fundamentally new.

For the purpose of keeping `alts_concurrent_connectivity_test` passing, my reason for preferring the existing workaround over increasing the handshake queue size to a very large value, is the latter workaround would IMO make the test less realistic of production situations than the former workaround by using an artificial value for the queue size.

Reply:
I should add that another approach could be to add some way to plumb down a bit to the ALTS handshake client to e.g. not use this queue, and then to e.g. set such a bit in the server-side ALTS creds within the `alts_concurrent_connectivity_test`. 

That would require some new plumbing to the ALTS creds though, so I leaned away from this for that reason, but I could also change this PR to do that if preferred.

Reply:
Understood. I meant that `g_client_handshake_queue` is not used when the code is running by the server side and vice versa. I agree that choosing a very large queue size is also hacky, and the current approach seems more reasonable.